In [1]:
import pandas as pd
import numpy as np
from datetime import date

# Cleaning grades_info Dataset

In [2]:
#Reading in the data for grades
grade_info = pd.read_csv("../raw_data/grade_info.csv")

In [3]:
grade_info.head(20)

grade sub_grade
0      B        B2
1      C        C4
2      C        C5
3      C        C1
4      B        B5
5      A        A4
6      E        E1
7      F        F2
8      C        C3
9      B        B1
10     D        D1
11     A        A1
12     B        B3
13     B        B4
14     C        C2
15     D        D2
16     A        A3
17     A        A5
18     D        D5
19     A        A2

In [4]:
#Getting dimensions
grade_info.shape

(36, 2)

In [5]:
#checking for missing values
grade_info.isna().sum()

grade        1
sub_grade    1
dtype: int64

In [6]:
grade_info.tail()

grade sub_grade
31     G        G2
32     G        G1
33     F        F5
34     G        G5
35   NaN       NaN

In [7]:
grade_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   grade      35 non-null     object
 1   sub_grade  35 non-null     object
dtypes: object(2)
memory usage: 704.0+ bytes


Removing the last row with missing values and saving as a new dataframe "grades_clean"

In [8]:
grades_clean = grade_info.drop([35]).copy()
grades_clean

grade sub_grade
0      B        B2
1      C        C4
2      C        C5
3      C        C1
4      B        B5
5      A        A4
6      E        E1
7      F        F2
8      C        C3
9      B        B1
10     D        D1
11     A        A1
12     B        B3
13     B        B4
14     C        C2
15     D        D2
16     A        A3
17     A        A5
18     D        D5
19     A        A2
20     E        E4
21     D        D3
22     D        D4
23     F        F3
24     E        E3
25     F        F4
26     F        F1
27     E        E5
28     G        G4
29     E        E2
30     G        G3
31     G        G2
32     G        G1
33     F        F5
34     G        G5

Writing grades_clean to csv and saving in clean_data files.

In [9]:
grades_clean.to_csv("../clean_data/grades_clean.csv")

### Reading in data dictionary.

In [10]:
data_dictionary = pd.read_csv("../raw_data/LCDataDictionary.csv")

Ensuring description column can be read in full.

In [11]:
pd.set_option("display.max_colwidth", None)
pd.set_option('display.max_rows', 120)

data_dictionary

LoanStatNew  \
0                    acc_now_delinq   
1              acc_open_past_24mths   
2                        addr_state   
3                          all_util   
4                        annual_inc   
5                  annual_inc_joint   
6                  application_type   
7                       avg_cur_bal   
8                    bc_open_to_buy   
9                           bc_util   
10         chargeoff_within_12_mths   
11          collection_recovery_fee   
12       collections_12_mths_ex_med   
13                      delinq_2yrs   
14                      delinq_amnt   
15                             desc   
16                              dti   
17                        dti_joint   
18                 earliest_cr_line   
19                       emp_length   
20                        emp_title   
21                  fico_range_high   
22                   fico_range_low   
23                      funded_amnt   
24                  funded_amnt_inv   
25                            grade   
26                   home_ownership   
27                               id   
28                          il_util   
29              initial_list_status   
30                           inq_fi   
31                     inq_last_12m   
32                   inq_last_6mths   
33                      installment   
34                         int_rate   
35                          issue_d   
36               last_credit_pull_d   
37             last_fico_range_high   
38              last_fico_range_low   
39                  last_pymnt_amnt   
40                     last_pymnt_d   
41                        loan_amnt   
42                      loan_status   
43                       max_bal_bc   
44                        member_id   
45               mo_sin_old_il_acct   
46             mo_sin_old_rev_tl_op   
47            mo_sin_rcnt_rev_tl_op   
48                   mo_sin_rcnt_tl   
49                         mort_acc   
50           mths_since_last_delinq   
51      mths_since_last_major_derog   
52           mths_since_last_record   
53               mths_since_rcnt_il   
54             mths_since_recent_bc   
55         mths_since_recent_bc_dlq   
56            mths_since_recent_inq   
57   mths_since_recent_revol_delinq   
58                     next_pymnt_d   
59            num_accts_ever_120_pd   
60                   num_actv_bc_tl   
61                  num_actv_rev_tl   
62                      num_bc_sats   
63                        num_bc_tl   
64                        num_il_tl   
65                    num_op_rev_tl   
66                    num_rev_accts   
67              num_rev_tl_bal_gt_0   
68                         num_sats   
69                 num_tl_120dpd_2m   
70                     num_tl_30dpd   
71               num_tl_90g_dpd_24m   
72               num_tl_op_past_12m   
73                         open_acc   
74                      open_acc_6m   
75                      open_il_12m   
76                      open_il_24m   
77                       open_il_6m   
78                      open_rv_12m   
79                      open_rv_24m   
80                        out_prncp   
81                    out_prncp_inv   
82                   pct_tl_nvr_dlq   
83                 percent_bc_gt_75   
84                      policy_code   
85                          pub_rec   
86             pub_rec_bankruptcies   
87                          purpose   
88                       pymnt_plan   
89                       recoveries   
90                        revol_bal   
91                       revol_util   
92                        sub_grade   
93                        tax_liens   
94                             term   
95                            title   
96                     tot_coll_amt   
97                      tot_cur_bal   
98                  tot_hi_cred_lim   
99                        total_acc   
100               total_bal_ex_mort   
101                    total_bal_il   
102    

### Viewing States dataset

In [12]:
#Reading in states dataset
states = pd.read_csv("../raw_data/state_names_info.csv")
states

state_abb      state_name
0         AL         Alabama
1         AK          Alaska
2         AZ         Arizona
3         AR        Arkansas
4         CA      California
5         CO        Colorado
6         CT     Connecticut
7         DE        Delaware
8         FL         Florida
9         GA         Georgia
10        HI          Hawaii
11        ID           Idaho
12        IL        Illinois
13        IN         Indiana
14        IA            Iowa
15        KS          Kansas
16        KY        Kentucky
17        LA       Louisiana
18        ME           Maine
19        MD        Maryland
20        MA   Massachusetts
21        MI        Michigan
22        MN       Minnesota
23        MS     Mississippi
24        MO        Missouri
25        MT         Montana
26        NE        Nebraska
27        NV          Nevada
28        NH   New Hampshire
29        NJ      New Jersey
30        NM      New Mexico
31        NY        New York
32        NC  North Carolina
33        ND    North Dakota
34        OH            Ohio
35        OK        Oklahoma
36        OR          Oregon
37        PA    Pennsylvania
38        RI    Rhode Island
39        SC  South Carolina
40        SD    South Dakota
41        TN       Tennessee
42        TX           Texas
43        UT            Utah
44        VT         Vermont
45        VA        Virginia
46        WA      Washington
47        WV   West Virginia
48        WI       Wisconsin
49        WY         Wyoming

# Cleaning lending_club_loans dataset

In [13]:
#Reading in the data for lending_club_loans
lending_club_loans = pd.read_csv("../raw_data/lending_club_loans.csv", low_memory = False)

In [14]:
lending_club_loans

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0      1077501.0  1296599.0     5000.0       5000.0           4975.0   
1      1077430.0  1314167.0     2500.0       2500.0           2500.0   
2      1077175.0  1313524.0     2400.0       2400.0           2400.0   
3      1076863.0  1277178.0    10000.0      10000.0          10000.0   
4      1075358.0  1311748.0     3000.0       3000.0           3000.0   
...          ...        ...        ...          ...              ...   
42533    72176.0    70868.0     2525.0       2525.0            225.0   
42534    71623.0    70735.0     6500.0       6500.0              0.0   
42535    70686.0    70681.0     5000.0       5000.0              0.0   
42536        NaN        NaN        NaN          NaN              NaN   
42537        NaN        NaN        NaN          NaN              NaN   

            term int_rate  installment sub_grade                 emp_title  \
0      36 months   10.65%       162.87        B2                       NaN   
1      60 months   15.27%        59.83        C4                     Ryder   
2      36 months   15.96%        84.33        C5                       NaN   
3      36 months   13.49%       339.31        C1       AIR RESOURCES BOARD   
4      60 months   12.69%        67.79        B5  University Medical Group   
...          ...      ...          ...       ...                       ...   
42533  36 months    9.33%        80.69        B3                       NaN   
42534  36 months    8.38%       204.84        A5                       NaN   
42535  36 months    7.75%       156.11        A3                 Homemaker   
42536        NaN      NaN          NaN       NaN                       NaN   
42537        NaN      NaN          NaN       NaN                       NaN   

       ... num_tl_90g_dpd_24m num_tl_op_past_12m  pct_tl_nvr_dlq  \
0      ...                NaN                NaN             NaN   
1      ...                NaN                NaN             NaN   
2      ...                NaN                NaN             NaN   
3      ...                NaN                NaN             NaN   
4      ...                NaN                NaN             NaN   
...    ...                ...                ...             ...   
42533  ...                NaN                NaN             NaN   
42534  ...                NaN                NaN             NaN   
42535  ...                NaN                NaN             NaN   
42536  ...                NaN                NaN             NaN   
42537  ...                NaN                NaN             NaN   

      percent_bc_gt_75 pub_rec_bankruptcies tax_liens tot_hi_cred_lim  \
0                  NaN                  0.0       0.0             NaN   
1                  NaN                  0.0       0.0             NaN   
2                  NaN                  0.0       0.0             NaN   
3                  NaN                  0.0       0.0             NaN   
4                  NaN                  0.0       0.0             NaN   
...                ...                  ...       ...             ...   
42533              NaN                  NaN       NaN             NaN   
42534              NaN                  NaN       NaN             NaN   
42535              NaN                  NaN       NaN             NaN   
42536              NaN                  NaN       NaN             NaN   
42537              NaN                  NaN       NaN             NaN   

      total_bal_ex_mort total_bc_limit total_il_high_credit_limit  
0                   NaN            NaN                        NaN  
1                   NaN            NaN                        NaN  
2                   NaN            NaN                        NaN  
3                   NaN            NaN                        NaN  
4                   NaN            NaN                        NaN  
...                 ...            ...                        ...  
42533               NaN            NaN    

In [15]:
lending_club_loans.shape

(42538, 114)

42538 rows and 114 columns. There are quite a few columns which will be irrelevant in predicting loan defaults so these will be removed.  

In [16]:
#Taking a look at all variable names
lending_club_loans.iloc[0]

id                                                                                                  1.0775e+06
member_id                                                                                           1.2966e+06
loan_amnt                                                                                                 5000
funded_amnt                                                                                               5000
funded_amnt_inv                                                                                           4975
term                                                                                                 36 months
int_rate                                                                                                10.65%
installment                                                                                             162.87
sub_grade                                                                                                   B2
e

To begin with, I will remove the following columns:- zip_code, id, member_id, url, emp_title & description.

In [17]:
loans_clean = lending_club_loans.drop(["id", "member_id", "zip_code", "url", 
                                       "emp_title", "desc"], axis = 1).copy()
loans_clean

loan_amnt  funded_amnt  funded_amnt_inv       term int_rate  \
0         5000.0       5000.0           4975.0  36 months   10.65%   
1         2500.0       2500.0           2500.0  60 months   15.27%   
2         2400.0       2400.0           2400.0  36 months   15.96%   
3        10000.0      10000.0          10000.0  36 months   13.49%   
4         3000.0       3000.0           3000.0  60 months   12.69%   
...          ...          ...              ...        ...      ...   
42533     2525.0       2525.0            225.0  36 months    9.33%   
42534     6500.0       6500.0              0.0  36 months    8.38%   
42535     5000.0       5000.0              0.0  36 months    7.75%   
42536        NaN          NaN              NaN        NaN      NaN   
42537        NaN          NaN              NaN        NaN      NaN   

       installment sub_grade emp_length home_ownership  annual_inc  ...  \
0           162.87        B2  10+ years           RENT     24000.0  ...   
1            59.83        C4   < 1 year           RENT     30000.0  ...   
2            84.33        C5  10+ years           RENT     12252.0  ...   
3           339.31        C1  10+ years           RENT     49200.0  ...   
4            67.79        B5     1 year           RENT     80000.0  ...   
...            ...       ...        ...            ...         ...  ...   
42533        80.69        B3   < 1 year           RENT    110000.0  ...   
42534       204.84        A5   < 1 year           NONE         NaN  ...   
42535       156.11        A3  10+ years       MORTGAGE     70000.0  ...   
42536          NaN       NaN        NaN            NaN         NaN  ...   
42537          NaN       NaN        NaN            NaN         NaN  ...   

      num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75  \
0                    NaN                NaN            NaN              NaN   
1                    NaN                NaN            NaN              NaN   
2                    NaN                NaN            NaN              NaN   
3                    NaN                NaN            NaN              NaN   
4                    NaN                NaN            NaN              NaN   
...                  ...                ...            ...              ...   
42533                NaN                NaN            NaN              NaN   
42534                NaN                NaN            NaN              NaN   
42535                NaN                NaN            NaN              NaN   
42536                NaN                NaN            NaN              NaN   
42537                NaN                NaN            NaN              NaN   

      pub_rec_bankruptcies tax_liens tot_hi_cred_lim  total_bal_ex_mort  \
0                      0.0       0.0             NaN                NaN   
1                      0.0       0.0             NaN                NaN   
2                      0.0       0.0             NaN                NaN   
3                      0.0       0.0             NaN                NaN   
4                      0.0       0.0             NaN                NaN   
...                    ...       ...             ...                ...   
42533                  NaN       NaN             NaN                NaN   
42534                  NaN       NaN             NaN                NaN   
42535                  NaN       NaN             NaN                NaN   
42536                  NaN       NaN             NaN                NaN   
42537                  NaN       NaN             NaN                NaN   

       total_bc_limit total_il_high_credit_limit  
0                 NaN                        NaN  
1                 NaN                        NaN  
2                 NaN                        NaN  
3                 NaN                        NaN  
4                 NaN                        NaN  
...               ...                        ...  
42533             NaN                        NaN  
42534  

In [18]:
loans_clean.dtypes

loan_amnt                         float64
funded_amnt                       float64
funded_amnt_inv                   float64
term                               object
int_rate                           object
installment                       float64
sub_grade                          object
emp_length                         object
home_ownership                     object
annual_inc                        float64
verification_status                object
issue_d                            object
loan_status                        object
pymnt_plan                         object
purpose                            object
title                              object
addr_state                         object
dti                               float64
delinq_2yrs                       float64
earliest_cr_line                   object
fico_range_low                    float64
fico_range_high                   float64
inq_last_6mths                    float64
mths_since_last_delinq            

Now taking a look at columns with missing values

In [19]:
loans_clean.isna().sum()

loan_amnt                             3
funded_amnt                           3
funded_amnt_inv                       3
term                                  3
int_rate                              3
installment                           3
sub_grade                             3
emp_length                         1115
home_ownership                        3
annual_inc                            7
verification_status                   3
issue_d                               3
loan_status                           3
pymnt_plan                            3
purpose                               3
title                                16
addr_state                            3
dti                                   3
delinq_2yrs                          32
earliest_cr_line                     32
fico_range_low                        3
fico_range_high                       3
inq_last_6mths                       32
mths_since_last_delinq            26929
mths_since_last_record            38887


There are numerous columns full of missing values which should be removed.

In [20]:
loans_clean = loans_clean.dropna(axis = "columns", how = "all")
loans_clean

loan_amnt  funded_amnt  funded_amnt_inv       term int_rate  \
0         5000.0       5000.0           4975.0  36 months   10.65%   
1         2500.0       2500.0           2500.0  60 months   15.27%   
2         2400.0       2400.0           2400.0  36 months   15.96%   
3        10000.0      10000.0          10000.0  36 months   13.49%   
4         3000.0       3000.0           3000.0  60 months   12.69%   
...          ...          ...              ...        ...      ...   
42533     2525.0       2525.0            225.0  36 months    9.33%   
42534     6500.0       6500.0              0.0  36 months    8.38%   
42535     5000.0       5000.0              0.0  36 months    7.75%   
42536        NaN          NaN              NaN        NaN      NaN   
42537        NaN          NaN              NaN        NaN      NaN   

       installment sub_grade emp_length home_ownership  annual_inc  ...  \
0           162.87        B2  10+ years           RENT     24000.0  ...   
1            59.83        C4   < 1 year           RENT     30000.0  ...   
2            84.33        C5  10+ years           RENT     12252.0  ...   
3           339.31        C1  10+ years           RENT     49200.0  ...   
4            67.79        B5     1 year           RENT     80000.0  ...   
...            ...       ...        ...            ...         ...  ...   
42533        80.69        B3   < 1 year           RENT    110000.0  ...   
42534       204.84        A5   < 1 year           NONE         NaN  ...   
42535       156.11        A3  10+ years       MORTGAGE     70000.0  ...   
42536          NaN       NaN        NaN            NaN         NaN  ...   
42537          NaN       NaN        NaN            NaN         NaN  ...   

      last_fico_range_high last_fico_range_low collections_12_mths_ex_med  \
0                    744.0               740.0                        0.0   
1                    499.0                 0.0                        0.0   
2                    719.0               715.0                        0.0   
3                    604.0               600.0                        0.0   
4                    694.0               690.0                        0.0   
...                    ...                 ...                        ...   
42533                714.0               710.0                        NaN   
42534                724.0               720.0                        NaN   
42535                794.0               790.0                        NaN   
42536                  NaN                 NaN                        NaN   
42537                  NaN                 NaN                        NaN   

      policy_code application_type acc_now_delinq chargeoff_within_12_mths  \
0             1.0       INDIVIDUAL            0.0                      0.0   
1             1.0       INDIVIDUAL            0.0                      0.0   
2             1.0       INDIVIDUAL            0.0                      0.0   
3             1.0       INDIVIDUAL            0.0                      0.0   
4             1.0       INDIVIDUAL            0.0                      0.0   
...           ...              ...            ...                      ...   
42533         1.0       INDIVIDUAL            NaN                      NaN   
42534         1.0       INDIVIDUAL            NaN                      NaN   
42535         1.0       INDIVIDUAL            NaN                      NaN   
42536         NaN              NaN            NaN                      NaN   
42537         NaN              NaN            NaN                      NaN   

       delinq_amnt  pub_rec_bankruptcies tax_liens  
0              0.0                   0.0       0.0  
1              0.0                   0.0       0.0  
2              0.0                   0.0       0.0  
3              0.0                   0.0       0.0  
4              0.0                   0.0       0.0  
...            ...                   ...       ...  
42533          NaN              

There are also rows which are full or missing values which should also be removed.

In [21]:
loans_clean = loans_clean.dropna(how = "all")
loans_clean

loan_amnt  funded_amnt  funded_amnt_inv       term int_rate  \
0         5000.0       5000.0           4975.0  36 months   10.65%   
1         2500.0       2500.0           2500.0  60 months   15.27%   
2         2400.0       2400.0           2400.0  36 months   15.96%   
3        10000.0      10000.0          10000.0  36 months   13.49%   
4         3000.0       3000.0           3000.0  60 months   12.69%   
...          ...          ...              ...        ...      ...   
42531     3500.0       3500.0            225.0  36 months   10.28%   
42532     1000.0       1000.0              0.0  36 months    9.64%   
42533     2525.0       2525.0            225.0  36 months    9.33%   
42534     6500.0       6500.0              0.0  36 months    8.38%   
42535     5000.0       5000.0              0.0  36 months    7.75%   

       installment sub_grade emp_length home_ownership  annual_inc  ...  \
0           162.87        B2  10+ years           RENT     24000.0  ...   
1            59.83        C4   < 1 year           RENT     30000.0  ...   
2            84.33        C5  10+ years           RENT     12252.0  ...   
3           339.31        C1  10+ years           RENT     49200.0  ...   
4            67.79        B5     1 year           RENT     80000.0  ...   
...            ...       ...        ...            ...         ...  ...   
42531       113.39        C1   < 1 year           RENT    180000.0  ...   
42532        32.11        B4   < 1 year           RENT     12000.0  ...   
42533        80.69        B3   < 1 year           RENT    110000.0  ...   
42534       204.84        A5   < 1 year           NONE         NaN  ...   
42535       156.11        A3  10+ years       MORTGAGE     70000.0  ...   

      last_fico_range_high last_fico_range_low collections_12_mths_ex_med  \
0                    744.0               740.0                        0.0   
1                    499.0                 0.0                        0.0   
2                    719.0               715.0                        0.0   
3                    604.0               600.0                        0.0   
4                    694.0               690.0                        0.0   
...                    ...                 ...                        ...   
42531                819.0               815.0                        NaN   
42532                784.0               780.0                        NaN   
42533                714.0               710.0                        NaN   
42534                724.0               720.0                        NaN   
42535                794.0               790.0                        NaN   

      policy_code application_type acc_now_delinq chargeoff_within_12_mths  \
0             1.0       INDIVIDUAL            0.0                      0.0   
1             1.0       INDIVIDUAL            0.0                      0.0   
2             1.0       INDIVIDUAL            0.0                      0.0   
3             1.0       INDIVIDUAL            0.0                      0.0   
4             1.0       INDIVIDUAL            0.0                      0.0   
...           ...              ...            ...                      ...   
42531         1.0       INDIVIDUAL            NaN                      NaN   
42532         1.0       INDIVIDUAL            NaN                      NaN   
42533         1.0       INDIVIDUAL            NaN                      NaN   
42534         1.0       INDIVIDUAL            NaN                      NaN   
42535         1.0       INDIVIDUAL            NaN                      NaN   

       delinq_amnt  pub_rec_bankruptcies tax_liens  
0              0.0                   0.0       0.0  
1              0.0                   0.0       0.0  
2              0.0                   0.0       0.0  
3              0.0                   0.0       0.0  
4              0.0                   0.0       0.0  
...            ...                   ...       ...  
42531          NaN              

In [22]:
loans_clean.iloc[0]

loan_amnt                            5000
funded_amnt                          5000
funded_amnt_inv                      4975
term                            36 months
int_rate                           10.65%
installment                        162.87
sub_grade                              B2
emp_length                      10+ years
home_ownership                       RENT
annual_inc                          24000
verification_status              Verified
issue_d                          Dec-2011
loan_status                    Fully Paid
pymnt_plan                              n
purpose                       credit_card
title                            Computer
addr_state                             AZ
dti                                 27.65
delinq_2yrs                             0
earliest_cr_line                 Jan-1985
fico_range_low                        735
fico_range_high                       739
inq_last_6mths                          1
mths_since_last_delinq            

In [23]:
loans_clean.shape

(42535, 54)

There are still quite a few columns which will be irrelevant in predicting loan defaults so these also need to be removed.  

mths_since_last_record, mths_since_last_delinq and next_pymnt_d also have thousands of missing values and are not relevant to predicting defaults so will be removed.

The following variables do not give insight into predicting defaults and are also being removed:
funded_amnt, funded_amnt_inv, pymnt_plan, last_fico_range_low, last_fico_range_high, collections_12_mths_ex_med, policy_code, delinq_amnt.

In [24]:
loans_clean = loans_clean.drop(["mths_since_last_record", "mths_since_last_delinq", "next_pymnt_d", 
                                "funded_amnt", "funded_amnt_inv", "pymnt_plan", 
                                "mths_since_last_delinq", "collections_12_mths_ex_med", 
                                "policy_code", "delinq_amnt", "last_fico_range_low", "last_fico_range_high"], 
                               axis = 1)
loans_clean

loan_amnt       term int_rate  installment sub_grade emp_length  \
0         5000.0  36 months   10.65%       162.87        B2  10+ years   
1         2500.0  60 months   15.27%        59.83        C4   < 1 year   
2         2400.0  36 months   15.96%        84.33        C5  10+ years   
3        10000.0  36 months   13.49%       339.31        C1  10+ years   
4         3000.0  60 months   12.69%        67.79        B5     1 year   
...          ...        ...      ...          ...       ...        ...   
42531     3500.0  36 months   10.28%       113.39        C1   < 1 year   
42532     1000.0  36 months    9.64%        32.11        B4   < 1 year   
42533     2525.0  36 months    9.33%        80.69        B3   < 1 year   
42534     6500.0  36 months    8.38%       204.84        A5   < 1 year   
42535     5000.0  36 months    7.75%       156.11        A3  10+ years   

      home_ownership  annual_inc verification_status   issue_d  ...  \
0               RENT     24000.0            Verified  Dec-2011  ...   
1               RENT     30000.0     Source Verified  Dec-2011  ...   
2               RENT     12252.0        Not Verified  Dec-2011  ...   
3               RENT     49200.0     Source Verified  Dec-2011  ...   
4               RENT     80000.0     Source Verified  Dec-2011  ...   
...              ...         ...                 ...       ...  ...   
42531           RENT    180000.0        Not Verified  Jun-2007  ...   
42532           RENT     12000.0        Not Verified  Jun-2007  ...   
42533           RENT    110000.0        Not Verified  Jun-2007  ...   
42534           NONE         NaN        Not Verified  Jun-2007  ...   
42535       MORTGAGE     70000.0        Not Verified  Jun-2007  ...   

      recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt  \
0           0.00                    0.00     Jan-2015          171.62   
1         117.08                    1.11     Apr-2013          119.66   
2           0.00                    0.00     Jun-2014          649.91   
3           0.00                    0.00     Jan-2015          357.48   
4           0.00                    0.00     Sep-2016           67.79   
...          ...                     ...          ...             ...   
42531       0.00                    0.00     Mar-2008            0.00   
42532       0.00                    0.00     Jun-2010           32.41   
42533       0.00                    0.00     Jun-2010           82.03   
42534       0.00                    0.00     Jun-2010          205.32   
42535       0.00                    0.00     Jun-2010          156.39   

       last_credit_pull_d  application_type acc_now_delinq  \
0                Sep-2016        INDIVIDUAL            0.0   
1                Sep-2016        INDIVIDUAL            0.0   
2                Sep-2016        INDIVIDUAL            0.0   
3                Apr-2016        INDIVIDUAL            0.0   
4                Sep-2016        INDIVIDUAL            0.0   
...                   ...               ...            ...   
42531            Feb-2013        INDIVIDUAL            NaN   
42532            Sep-2014        INDIVIDUAL            NaN   
42533            May-2007        INDIVIDUAL            NaN   
42534            Aug-2007        INDIVIDUAL            NaN   
42535            Feb-2015        INDIVIDUAL            NaN   

       chargeoff_within_12_mths  pub_rec_bankruptcies  tax_liens  
0                           0.0                   0.0        0.0  
1                           0.0                   0.0        0.0  
2                           0.0                   0.0        0.0  
3                           0.0                   0.0        0.0  
4                           0.0                   0.0        0.0  
...                         ...                   ...        ...  
42531                       NaN                   NaN        NaN  
42532                       NaN                   NaN        NaN  
42533                       NaN         

In [25]:
loans_clean.isna().sum()

loan_amnt                      0
term                           0
int_rate                       0
installment                    0
sub_grade                      0
emp_length                  1112
home_ownership                 0
annual_inc                     4
verification_status            0
issue_d                        0
loan_status                    0
purpose                        0
title                         13
addr_state                     0
dti                            0
delinq_2yrs                   29
earliest_cr_line              29
fico_range_low                 0
fico_range_high                0
inq_last_6mths                29
open_acc                      29
pub_rec                       29
revol_bal                      0
revol_util                    90
total_acc                     29
initial_list_status            0
out_prncp                      0
out_prncp_inv                  0
total_pymnt                    0
total_pymnt_inv                0
total_rec_

In [26]:
loans_clean.shape

(42535, 43)

In [27]:
#Removing further variables which are irrelevant in default prediction
loans_clean = loans_clean.drop(["tax_liens", "title", "inq_last_6mths", "pub_rec", 
                                "pub_rec_bankruptcies", "revol_bal", "revol_util", 
                                "initial_list_status", "out_prncp", "out_prncp_inv", "total_pymnt",
                                "total_pymnt_inv", "total_rec_prncp", "total_rec_int", "total_rec_late_fee",
                                "recoveries", "collection_recovery_fee", "last_pymnt_d", "last_pymnt_amnt",
                                "last_credit_pull_d", "chargeoff_within_12_mths"], axis = 1)
loans_clean

loan_amnt       term int_rate  installment sub_grade emp_length  \
0         5000.0  36 months   10.65%       162.87        B2  10+ years   
1         2500.0  60 months   15.27%        59.83        C4   < 1 year   
2         2400.0  36 months   15.96%        84.33        C5  10+ years   
3        10000.0  36 months   13.49%       339.31        C1  10+ years   
4         3000.0  60 months   12.69%        67.79        B5     1 year   
...          ...        ...      ...          ...       ...        ...   
42531     3500.0  36 months   10.28%       113.39        C1   < 1 year   
42532     1000.0  36 months    9.64%        32.11        B4   < 1 year   
42533     2525.0  36 months    9.33%        80.69        B3   < 1 year   
42534     6500.0  36 months    8.38%       204.84        A5   < 1 year   
42535     5000.0  36 months    7.75%       156.11        A3  10+ years   

      home_ownership  annual_inc verification_status   issue_d  ...  \
0               RENT     24000.0            Verified  Dec-2011  ...   
1               RENT     30000.0     Source Verified  Dec-2011  ...   
2               RENT     12252.0        Not Verified  Dec-2011  ...   
3               RENT     49200.0     Source Verified  Dec-2011  ...   
4               RENT     80000.0     Source Verified  Dec-2011  ...   
...              ...         ...                 ...       ...  ...   
42531           RENT    180000.0        Not Verified  Jun-2007  ...   
42532           RENT     12000.0        Not Verified  Jun-2007  ...   
42533           RENT    110000.0        Not Verified  Jun-2007  ...   
42534           NONE         NaN        Not Verified  Jun-2007  ...   
42535       MORTGAGE     70000.0        Not Verified  Jun-2007  ...   

      addr_state    dti delinq_2yrs  earliest_cr_line  fico_range_low  \
0             AZ  27.65         0.0          Jan-1985           735.0   
1             GA   1.00         0.0          Apr-1999           740.0   
2             IL   8.72         0.0          Nov-2001           735.0   
3             CA  20.00         0.0          Feb-1996           690.0   
4             OR  17.94         0.0          Jan-1996           695.0   
...          ...    ...         ...               ...             ...   
42531         NY  10.00         NaN               NaN           685.0   
42532         MA  10.00         NaN               NaN           695.0   
42533         NY  10.00         NaN               NaN           710.0   
42534         NY   4.00         NaN               NaN           740.0   
42535         CT   8.81         NaN               NaN           770.0   

      fico_range_high  open_acc  total_acc  application_type  acc_now_delinq  
0               739.0       3.0        9.0        INDIVIDUAL             0.0  
1               744.0       3.0        4.0        INDIVIDUAL             0.0  
2               739.0       2.0       10.0        INDIVIDUAL             0.0  
3               694.0      10.0       37.0        INDIVIDUAL             0.0  
4               699.0      15.0       38.0        INDIVIDUAL             0.0  
...               ...       ...        ...               ...             ...  
42531           689.0       NaN        NaN        INDIVIDUAL             NaN  
42532           699.0       NaN        NaN        INDIVIDUAL             NaN  
42533           714.0       NaN        NaN        INDIVIDUAL             NaN  
42534           744.0       NaN        NaN        INDIVIDUAL             NaN  
42535           774.0       NaN        NaN        INDIVIDUAL             NaN  

[42535 rows x 22 columns]

In [28]:
loans_clean.isna().sum()

loan_amnt                 0
term                      0
int_rate                  0
installment               0
sub_grade                 0
emp_length             1112
home_ownership            0
annual_inc                4
verification_status       0
issue_d                   0
loan_status               0
purpose                   0
addr_state                0
dti                       0
delinq_2yrs              29
earliest_cr_line         29
fico_range_low            0
fico_range_high           0
open_acc                 29
total_acc                29
application_type          0
acc_now_delinq           29
dtype: int64

### Missing Values

In [29]:
#Employment length has over 1000 missing values.
#Obtaining a list of values for this variable
loans_clean['emp_length'].unique()

array(['10+ years', '< 1 year', '1 year', '3 years', '8 years', '9 years',
       '4 years', '5 years', '6 years', '2 years', '7 years', nan],
      dtype=object)

In [30]:
loans_clean.dtypes

loan_amnt              float64
term                    object
int_rate                object
installment            float64
sub_grade               object
emp_length              object
home_ownership          object
annual_inc             float64
verification_status     object
issue_d                 object
loan_status             object
purpose                 object
addr_state              object
dti                    float64
delinq_2yrs            float64
earliest_cr_line        object
fico_range_low         float64
fico_range_high        float64
open_acc               float64
total_acc              float64
application_type        object
acc_now_delinq         float64
dtype: object

In [31]:
#Removing missing values in emp_length.
loans_clean = loans_clean.dropna(how = "any", subset = ["emp_length"])

Binning values in emp_length. 

In [32]:
def emp_length_num (row):
    if row['emp_length'] in ["< 1 year", "1 year", ]:
        return '< 2 years'
    if row['emp_length'] in ["2 years", "3 years", "4 years"]:
        return '2 - 4 years'
    if row['emp_length'] in [ "5 years", "6 years", "7 years", "8 years", "9 years"]:
        return '5 - 9 years'
    return '10+ years'

loans_clean['emp_length'] = loans_clean.apply(lambda row: emp_length_num(row), axis=1)

loans_clean["emp_length"].value_counts()

<ipython-input-32-1bf02f197af0>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  loans_clean['emp_length'] = loans_clean.apply(lambda row: emp_length_num(row), axis=1)


2 - 4 years    12756
5 - 9 years    10641
10+ years       9369
< 2 years       8657
Name: emp_length, dtype: int64

In [33]:
#Verifying the employment length column has been changed
loans_clean['emp_length'].value_counts()

2 - 4 years    12756
5 - 9 years    10641
10+ years       9369
< 2 years       8657
Name: emp_length, dtype: int64

Removing the rows with missing values for the following columns:

* annual_inc
* earliest_cr_line
* open_acc
* total_acc

In [34]:
loans_clean = loans_clean.dropna(how = "any", subset = ["annual_inc",
                                                       "earliest_cr_line",
                                                       "open_acc",
                                                       "total_acc"])

In [35]:
loans_clean.isna().sum()

loan_amnt              0
term                   0
int_rate               0
installment            0
sub_grade              0
emp_length             0
home_ownership         0
annual_inc             0
verification_status    0
issue_d                0
loan_status            0
purpose                0
addr_state             0
dti                    0
delinq_2yrs            0
earliest_cr_line       0
fico_range_low         0
fico_range_high        0
open_acc               0
total_acc              0
application_type       0
acc_now_delinq         0
dtype: int64

# Joining all three datsets together

Firstly, joining loans dataset with states dataset to show state abbreviations as names.

In [36]:
loans_data_clean = pd.merge(loans_clean, states, left_on="addr_state", right_on="state_abb")
#removing the columns "add_state" and "state_abb" as these are now replaced with "state_name"
loans_data_clean = loans_data_clean.drop(["addr_state", "state_abb"], axis = 1)
loans_data_clean

loan_amnt       term int_rate  installment sub_grade   emp_length  \
0         5000.0  36 months   10.65%       162.87        B2    10+ years   
1         5000.0  36 months    7.90%       156.46        A4  2 - 4 years   
2         6500.0  60 months   14.65%       153.45        C3  5 - 9 years   
3         6000.0  36 months   12.42%       200.50        B4    < 2 years   
4        12000.0  36 months   16.29%       423.61        D1  5 - 9 years   
...          ...        ...      ...          ...       ...          ...   
41165     1000.0  36 months   12.49%        33.45        D3    < 2 years   
41166     2600.0  36 months   14.38%        89.35        E4    10+ years   
41167     3000.0  36 months    9.33%        95.87        B3  2 - 4 years   
41168     5000.0  36 months   10.59%       162.73        C2    < 2 years   
41169     1200.0  36 months   11.54%        39.60        C5    < 2 years   

      home_ownership  annual_inc verification_status   issue_d  ...    dti  \
0               RENT     24000.0            Verified  Dec-2011  ...  27.65   
1               RENT     36000.0     Source Verified  Dec-2011  ...  11.20   
2                OWN     72000.0        Not Verified  Dec-2011  ...  16.12   
3               RENT     36852.0     Source Verified  Dec-2011  ...  10.62   
4               RENT     88365.0            Verified  Dec-2011  ...  16.85   
...              ...         ...                 ...       ...  ...    ...   
41165           RENT     32000.0        Not Verified  Jul-2007  ...   5.21   
41166            OWN     42000.0        Not Verified  Jul-2007  ...  16.91   
41167           RENT     36600.0        Not Verified  Dec-2007  ...   2.69   
41168           RENT     15000.0        Not Verified  Oct-2007  ...  24.40   
41169           RENT     20000.0        Not Verified  Jul-2007  ...   2.04   

      delinq_2yrs  earliest_cr_line  fico_range_low fico_range_high  open_acc  \
0             0.0          Jan-1985           735.0           739.0       3.0   
1             0.0          Nov-2004           730.0           734.0       9.0   
2             0.0          Jan-1998           695.0           699.0      14.0   
3             0.0          Apr-2007           695.0           699.0       7.0   
4             0.0          Oct-1996           680.0           684.0       6.0   
...           ...               ...             ...             ...       ...   
41165         4.0          Oct-2000           650.0           654.0      12.0   
41166         4.0          Apr-1992           640.0           644.0      14.0   
41167         0.0          Feb-1995           695.0           699.0       4.0   
41168         0.0          Jan-1999           735.0           739.0       6.0   
41169         0.0          May-2004           660.0           664.0       3.0   

       total_acc  application_type  acc_now_delinq state_name  
0            9.0        INDIVIDUAL             0.0    Arizona  
1           12.0        INDIVIDUAL             0.0    Arizona  
2           23.0        INDIVIDUAL             0.0    Arizona  
3           12.0        INDIVIDUAL             0.0    Arizona  
4           15.0        INDIVIDUAL             0.0    Arizona  
...          ...               ...             ...        ...  
41165       27.0        INDIVIDUAL             0.0    Indiana  
41166       28.0        INDIVIDUAL             0.0    Indiana  
41167        5.0        INDIVIDUAL             0.0      Maine  
41168        6.0        INDIVIDUAL             0.0      Maine  
41169        4.0        INDIVIDUAL             0.0      Maine  

[41170 rows x 22 columns]

Now joining loans_data with grades dataset.

In [37]:
loans_data_clean = pd.merge(loans_data_clean, grades_clean, how = "left")
loans_data_clean

loan_amnt       term int_rate  installment sub_grade   emp_length  \
0         5000.0  36 months   10.65%       162.87        B2    10+ years   
1         5000.0  36 months    7.90%       156.46        A4  2 - 4 years   
2         6500.0  60 months   14.65%       153.45        C3  5 - 9 years   
3         6000.0  36 months   12.42%       200.50        B4    < 2 years   
4        12000.0  36 months   16.29%       423.61        D1  5 - 9 years   
...          ...        ...      ...          ...       ...          ...   
41165     1000.0  36 months   12.49%        33.45        D3    < 2 years   
41166     2600.0  36 months   14.38%        89.35        E4    10+ years   
41167     3000.0  36 months    9.33%        95.87        B3  2 - 4 years   
41168     5000.0  36 months   10.59%       162.73        C2    < 2 years   
41169     1200.0  36 months   11.54%        39.60        C5    < 2 years   

      home_ownership  annual_inc verification_status   issue_d  ...  \
0               RENT     24000.0            Verified  Dec-2011  ...   
1               RENT     36000.0     Source Verified  Dec-2011  ...   
2                OWN     72000.0        Not Verified  Dec-2011  ...   
3               RENT     36852.0     Source Verified  Dec-2011  ...   
4               RENT     88365.0            Verified  Dec-2011  ...   
...              ...         ...                 ...       ...  ...   
41165           RENT     32000.0        Not Verified  Jul-2007  ...   
41166            OWN     42000.0        Not Verified  Jul-2007  ...   
41167           RENT     36600.0        Not Verified  Dec-2007  ...   
41168           RENT     15000.0        Not Verified  Oct-2007  ...   
41169           RENT     20000.0        Not Verified  Jul-2007  ...   

      delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high open_acc  \
0             0.0         Jan-1985           735.0            739.0      3.0   
1             0.0         Nov-2004           730.0            734.0      9.0   
2             0.0         Jan-1998           695.0            699.0     14.0   
3             0.0         Apr-2007           695.0            699.0      7.0   
4             0.0         Oct-1996           680.0            684.0      6.0   
...           ...              ...             ...              ...      ...   
41165         4.0         Oct-2000           650.0            654.0     12.0   
41166         4.0         Apr-1992           640.0            644.0     14.0   
41167         0.0         Feb-1995           695.0            699.0      4.0   
41168         0.0         Jan-1999           735.0            739.0      6.0   
41169         0.0         May-2004           660.0            664.0      3.0   

       total_acc  application_type  acc_now_delinq  state_name grade  
0            9.0        INDIVIDUAL             0.0     Arizona     B  
1           12.0        INDIVIDUAL             0.0     Arizona     A  
2           23.0        INDIVIDUAL             0.0     Arizona     C  
3           12.0        INDIVIDUAL             0.0     Arizona     B  
4           15.0        INDIVIDUAL             0.0     Arizona     D  
...          ...               ...             ...         ...   ...  
41165       27.0        INDIVIDUAL             0.0     Indiana     D  
41166       28.0        INDIVIDUAL             0.0     Indiana     E  
41167        5.0        INDIVIDUAL             0.0       Maine     B  
41168        6.0        INDIVIDUAL             0.0       Maine     C  
41169        4.0        INDIVIDUAL             0.0       Maine     C  

[41170 rows x 23 columns]

Removing the column "sub_grade" as this has now replaced with "grade"

In [38]:
loans_data_clean = loans_data_clean.drop(["sub_grade"], axis = 1)
loans_data_clean

loan_amnt       term int_rate  installment   emp_length home_ownership  \
0         5000.0  36 months   10.65%       162.87    10+ years           RENT   
1         5000.0  36 months    7.90%       156.46  2 - 4 years           RENT   
2         6500.0  60 months   14.65%       153.45  5 - 9 years            OWN   
3         6000.0  36 months   12.42%       200.50    < 2 years           RENT   
4        12000.0  36 months   16.29%       423.61  5 - 9 years           RENT   
...          ...        ...      ...          ...          ...            ...   
41165     1000.0  36 months   12.49%        33.45    < 2 years           RENT   
41166     2600.0  36 months   14.38%        89.35    10+ years            OWN   
41167     3000.0  36 months    9.33%        95.87  2 - 4 years           RENT   
41168     5000.0  36 months   10.59%       162.73    < 2 years           RENT   
41169     1200.0  36 months   11.54%        39.60    < 2 years           RENT   

       annual_inc verification_status   issue_d  \
0         24000.0            Verified  Dec-2011   
1         36000.0     Source Verified  Dec-2011   
2         72000.0        Not Verified  Dec-2011   
3         36852.0     Source Verified  Dec-2011   
4         88365.0            Verified  Dec-2011   
...           ...                 ...       ...   
41165     32000.0        Not Verified  Jul-2007   
41166     42000.0        Not Verified  Jul-2007   
41167     36600.0        Not Verified  Dec-2007   
41168     15000.0        Not Verified  Oct-2007   
41169     20000.0        Not Verified  Jul-2007   

                                               loan_status  ... delinq_2yrs  \
0                                               Fully Paid  ...         0.0   
1                                               Fully Paid  ...         0.0   
2                                               Fully Paid  ...         0.0   
3                                               Fully Paid  ...         0.0   
4                                               Fully Paid  ...         0.0   
...                                                    ...  ...         ...   
41165   Does not meet the credit policy. Status:Fully Paid  ...         4.0   
41166  Does not meet the credit policy. Status:Charged Off  ...         4.0   
41167                                           Fully Paid  ...         0.0   
41168                                           Fully Paid  ...         0.0   
41169                                           Fully Paid  ...         0.0   

       earliest_cr_line  fico_range_low fico_range_high  open_acc  total_acc  \
0              Jan-1985           735.0           739.0       3.0        9.0   
1              Nov-2004           730.0           734.0       9.0       12.0   
2              Jan-1998           695.0           699.0      14.0       23.0   
3              Apr-2007           695.0           699.0       7.0       12.0   
4              Oct-1996           680.0           684.0       6.0       15.0   
...                 ...             ...             ...       ...        ...   
41165          Oct-2000           650.0           654.0      12.0       27.0   
41166          Apr-1992           640.0           644.0      14.0       28.0   
41167          Feb-1995           695.0           699.0       4.0        5.0   
41168          Jan-1999           735.0           739.0       6.0        6.0   
41169          May-2004           660.0           664.0       3.0        4.0   

       application_type  acc_now_delinq state_name  grade  
0            INDIVIDUAL             0.0    Arizona      B  
1            INDIVIDUAL             0.0    Arizona      A  
2            INDIVIDUAL             0.0    Arizona      C  
3            INDIVIDUAL             0.0    Arizona      B  
4            INDIVIDUAL             0.0    Arizona      D  
...                 ...             ...        ...    ...  
41165        INDIVIDUAL             0.0    Indiana      D  
41166        INDIVIDUAL        

# Target Variable - "loan_status"

Taking a closer look at the "loan_status" variable which will be the target variable used to predict defaults.

In [39]:
loans_data_clean['loan_status'].unique()

array(['Fully Paid', 'Charged Off', 'Current', 'Late (16-30 days)',
       'Does not meet the credit policy. Status:Fully Paid',
       'Does not meet the credit policy. Status:Charged Off',
       'In Grace Period', 'Late (31-120 days)', 'Default'], dtype=object)

There are a list of 9 options under "loan_status", however whether a loan has been fully paid or charged off are the important values.

Taking a look at how many rows there are for each of the 9 options.

In [40]:
loans_data_clean['loan_status'].value_counts()

Fully Paid                                             32561
Charged Off                                             5408
Does not meet the credit policy. Status:Fully Paid      1935
Does not meet the credit policy. Status:Charged Off      741
Current                                                  493
In Grace Period                                           15
Late (31-120 days)                                        12
Late (16-30 days)                                          4
Default                                                    1
Name: loan_status, dtype: int64

There are applications which show "Does not meet the credit policy." This does not look great and should not be used in predicting future defaults. These rows will be removed.

In [41]:
loans_data_clean = loans_data_clean[-loans_data_clean["loan_status"].isin(["Does not meet the credit policy. Status:Fully Paid", "Does not meet the credit policy. Status:Charged Off"])]

In [42]:
loans_data_clean['loan_status'].value_counts()

Fully Paid            32561
Charged Off            5408
Current                 493
In Grace Period          15
Late (31-120 days)       12
Late (16-30 days)         4
Default                   1
Name: loan_status, dtype: int64

### Checking other Variables

In [43]:
loans_data_clean["home_ownership"].unique()

array(['RENT', 'OWN', 'MORTGAGE', 'OTHER', 'NONE'], dtype=object)

The home_ownership variable has values showing "other" and "none". The data dictionary shows that LendingClub only record the values "rent", "own", "mortgage" or "other" so I will remove rows which show "none".

In [44]:
loans_data_clean = loans_data_clean[-loans_data_clean["home_ownership"].str.contains("NONE")]
loans_data_clean["home_ownership"].unique()

array(['RENT', 'OWN', 'MORTGAGE', 'OTHER'], dtype=object)

In [45]:
loans_data_clean["verification_status"].unique()

array(['Verified', 'Source Verified', 'Not Verified'], dtype=object)

In [46]:
loans_data_clean["purpose"].unique()

array(['credit_card', 'wedding', 'debt_consolidation', 'other',
       'small_business', 'major_purchase', 'home_improvement', 'moving',
       'car', 'house', 'medical', 'vacation', 'renewable_energy',
       'educational'], dtype=object)

In [47]:
loans_data_clean["application_type"].unique()

array(['INDIVIDUAL'], dtype=object)

In [48]:
lending_club_loans["application_type"].unique()

array(['INDIVIDUAL', nan], dtype=object)

After comparing the "application_type" variable in the raw data and clean data, the application type is only either "individual" or "nan". It would appear that this column doesn't add much value so I will lose it.

In [49]:
loans_data_clean = loans_data_clean.drop(["application_type"], axis = 1)

In [50]:
loans_data_clean["issue_d"].unique()

array(['Dec-2011', 'Nov-2011', 'Oct-2011', 'Sep-2011', 'Aug-2011',
       'Jul-2011', 'Jun-2011', 'May-2011', 'Apr-2011', 'Mar-2011',
       'Feb-2011', 'Jan-2011', 'Dec-2010', 'Nov-2010', 'Oct-2010',
       'Sep-2010', 'Aug-2010', 'Jul-2010', 'Jun-2010', 'May-2010',
       'Apr-2010', 'Mar-2010', 'Feb-2010', 'Jan-2010', 'Dec-2009',
       'Nov-2009', 'Oct-2009', 'Sep-2009', 'Aug-2009', 'Jul-2009',
       'Jun-2009', 'May-2009', 'Apr-2009', 'Mar-2009', 'Feb-2009',
       'Jan-2009', 'Dec-2008', 'Nov-2008', 'Oct-2008', 'Sep-2008',
       'Aug-2008', 'Jul-2008', 'Jun-2008', 'May-2008', 'Apr-2008',
       'Mar-2008', 'Feb-2008', 'Jan-2008', 'Dec-2007', 'Nov-2007',
       'Oct-2007', 'Aug-2007', 'Jul-2007', 'Sep-2007', 'Jun-2007'],
      dtype=object)

In [52]:
loans_data_clean["acc_now_delinq"].unique()

array([0.])

The variable "acc_now_delinq" is full of 0 values so won't add any value in analysing the data so will also be removed.

In [53]:
loans_data_clean = loans_data_clean.drop(["acc_now_delinq"], axis = 1)
loans_data_clean

loan_amnt       term int_rate  installment   emp_length home_ownership  \
0         5000.0  36 months   10.65%       162.87    10+ years           RENT   
1         5000.0  36 months    7.90%       156.46  2 - 4 years           RENT   
2         6500.0  60 months   14.65%       153.45  5 - 9 years            OWN   
3         6000.0  36 months   12.42%       200.50    < 2 years           RENT   
4        12000.0  36 months   16.29%       423.61  5 - 9 years           RENT   
...          ...        ...      ...          ...          ...            ...   
41155     5100.0  36 months    8.38%       160.72  2 - 4 years           RENT   
41156     5300.0  36 months    8.38%       167.02  5 - 9 years       MORTGAGE   
41167     3000.0  36 months    9.33%        95.87  2 - 4 years           RENT   
41168     5000.0  36 months   10.59%       162.73    < 2 years           RENT   
41169     1200.0  36 months   11.54%        39.60    < 2 years           RENT   

       annual_inc verification_status   issue_d loan_status  \
0         24000.0            Verified  Dec-2011  Fully Paid   
1         36000.0     Source Verified  Dec-2011  Fully Paid   
2         72000.0        Not Verified  Dec-2011  Fully Paid   
3         36852.0     Source Verified  Dec-2011  Fully Paid   
4         88365.0            Verified  Dec-2011  Fully Paid   
...           ...                 ...       ...         ...   
41155     32000.0        Not Verified  Jul-2007  Fully Paid   
41156     74000.0        Not Verified  Jul-2007  Fully Paid   
41167     36600.0        Not Verified  Dec-2007  Fully Paid   
41168     15000.0        Not Verified  Oct-2007  Fully Paid   
41169     20000.0        Not Verified  Jul-2007  Fully Paid   

                  purpose    dti  delinq_2yrs earliest_cr_line  \
0             credit_card  27.65          0.0         Jan-1985   
1                 wedding  11.20          0.0         Nov-2004   
2      debt_consolidation  16.12          0.0         Jan-1998   
3      debt_consolidation  10.62          0.0         Apr-2007   
4      debt_consolidation  16.85          0.0         Oct-1996   
...                   ...    ...          ...              ...   
41155            vacation  14.78          0.0         Jun-1995   
41156         credit_card  14.37          0.0         Feb-1992   
41167               other   2.69          0.0         Feb-1995   
41168               other  24.40          0.0         Jan-1999   
41169         credit_card   2.04          0.0         May-2004   

       fico_range_low  fico_range_high  open_acc  total_acc state_name grade  
0               735.0            739.0       3.0        9.0    Arizona     B  
1               730.0            734.0       9.0       12.0    Arizona     A  
2               695.0            699.0      14.0       23.0    Arizona     C  
3               695.0            699.0       7.0       12.0    Arizona     B  
4               680.0            684.0       6.0       15.0    Arizona     D  
...               ...              ...       ...        ...        ...   ...  
41155           770.0            774.0      11.0       22.0    Indiana     A  
41156           795.0            799.0      15.0       29.0    Indiana     A  
41167           695.0            699.0       4.0        5.0      Maine     B  
41168           735.0            739.0       6.0        6.0      Maine     C  
41169           660.0            664.0       3.0        4.0      Maine     C  

[38491 rows x 20 columns]

In [54]:
loans_data_clean["delinq_2yrs"].unique()

array([ 0.,  1.,  3.,  2.,  4.,  6., 11.,  5.,  7.,  8.,  9.])

Removing "%" from interest rate column

In [55]:
loans_data_clean['int_rate'] = loans_data_clean['int_rate'].str.replace('%', '')

In [56]:
loans_data_clean["term"].unique()

array(['36 months', '60 months'], dtype=object)

In [57]:
loans_data_clean["fico_range_high"].unique()

array([739., 734., 699., 684., 694., 679., 674., 689., 719., 794., 729.,
       709., 714., 669., 759., 774., 744., 754., 704., 724., 664., 784.,
       769., 749., 779., 799., 789., 764., 804., 809., 814., 819., 824.,
       829., 629., 634.])

In [58]:
loans_data_clean["fico_range_low"].unique()

array([735., 730., 695., 680., 690., 675., 670., 685., 715., 790., 725.,
       705., 710., 665., 755., 770., 740., 750., 700., 720., 660., 780.,
       765., 745., 775., 795., 785., 760., 800., 805., 810., 815., 820.,
       825., 625., 630.])

The variable "fico_range_low" is 4 points off the "fico_range_high" score. Will take the average of the two scores and hold in a new column.

In [59]:
loans_data_clean["fico_range"] = (loans_data_clean["fico_range_low"] + loans_data_clean["fico_range_high"])/2

In [60]:
loans_data_clean = loans_data_clean.drop(["fico_range_low", "fico_range_high"], axis = 1)
loans_data_clean

loan_amnt       term int_rate  installment   emp_length home_ownership  \
0         5000.0  36 months    10.65       162.87    10+ years           RENT   
1         5000.0  36 months     7.90       156.46  2 - 4 years           RENT   
2         6500.0  60 months    14.65       153.45  5 - 9 years            OWN   
3         6000.0  36 months    12.42       200.50    < 2 years           RENT   
4        12000.0  36 months    16.29       423.61  5 - 9 years           RENT   
...          ...        ...      ...          ...          ...            ...   
41155     5100.0  36 months     8.38       160.72  2 - 4 years           RENT   
41156     5300.0  36 months     8.38       167.02  5 - 9 years       MORTGAGE   
41167     3000.0  36 months     9.33        95.87  2 - 4 years           RENT   
41168     5000.0  36 months    10.59       162.73    < 2 years           RENT   
41169     1200.0  36 months    11.54        39.60    < 2 years           RENT   

       annual_inc verification_status   issue_d loan_status  \
0         24000.0            Verified  Dec-2011  Fully Paid   
1         36000.0     Source Verified  Dec-2011  Fully Paid   
2         72000.0        Not Verified  Dec-2011  Fully Paid   
3         36852.0     Source Verified  Dec-2011  Fully Paid   
4         88365.0            Verified  Dec-2011  Fully Paid   
...           ...                 ...       ...         ...   
41155     32000.0        Not Verified  Jul-2007  Fully Paid   
41156     74000.0        Not Verified  Jul-2007  Fully Paid   
41167     36600.0        Not Verified  Dec-2007  Fully Paid   
41168     15000.0        Not Verified  Oct-2007  Fully Paid   
41169     20000.0        Not Verified  Jul-2007  Fully Paid   

                  purpose    dti  delinq_2yrs earliest_cr_line  open_acc  \
0             credit_card  27.65          0.0         Jan-1985       3.0   
1                 wedding  11.20          0.0         Nov-2004       9.0   
2      debt_consolidation  16.12          0.0         Jan-1998      14.0   
3      debt_consolidation  10.62          0.0         Apr-2007       7.0   
4      debt_consolidation  16.85          0.0         Oct-1996       6.0   
...                   ...    ...          ...              ...       ...   
41155            vacation  14.78          0.0         Jun-1995      11.0   
41156         credit_card  14.37          0.0         Feb-1992      15.0   
41167               other   2.69          0.0         Feb-1995       4.0   
41168               other  24.40          0.0         Jan-1999       6.0   
41169         credit_card   2.04          0.0         May-2004       3.0   

       total_acc state_name grade  fico_range  
0            9.0    Arizona     B       737.0  
1           12.0    Arizona     A       732.0  
2           23.0    Arizona     C       697.0  
3           12.0    Arizona     B       697.0  
4           15.0    Arizona     D       682.0  
...          ...        ...   ...         ...  
41155       22.0    Indiana     A       772.0  
41156       29.0    Indiana     A       797.0  
41167        5.0      Maine     B       697.0  
41168        6.0      Maine     C       737.0  
41169        4.0      Maine     C       662.0  

[38491 rows x 19 columns]

Changing the "fico_range" values to categorical.

In [61]:
bins = [0, 580, 670, 740, 800, np.inf]
names = ["Poor", "Fair", "Good", "Very Good", "Exceptional"]

loans_data_clean['credit_score'] = pd.cut(loans_data_clean['fico_range'], bins, labels=names)
loans_data_clean.head(50)

loan_amnt       term int_rate  installment   emp_length home_ownership  \
0      5000.0  36 months    10.65       162.87    10+ years           RENT   
1      5000.0  36 months     7.90       156.46  2 - 4 years           RENT   
2      6500.0  60 months    14.65       153.45  5 - 9 years            OWN   
3      6000.0  36 months    12.42       200.50    < 2 years           RENT   
4     12000.0  36 months    16.29       423.61  5 - 9 years           RENT   
5      6000.0  36 months    13.49       203.59  5 - 9 years           RENT   
6      5600.0  36 months    14.65       193.17    < 2 years       MORTGAGE   
7      7400.0  36 months    13.49       251.09    < 2 years            OWN   
8     24250.0  60 months    23.91       519.76  5 - 9 years       MORTGAGE   
9      2500.0  36 months    11.71        82.69  5 - 9 years       MORTGAGE   
10    12000.0  36 months     8.90       381.04  5 - 9 years       MORTGAGE   
11    18000.0  60 months    11.71       397.77    < 2 years           RENT   
12     7200.0  36 months     6.03       219.14  5 - 9 years       MORTGAGE   
13     4200.0  60 months    17.27       105.00  5 - 9 years           RENT   
14    20000.0  60 months    18.64       514.86  2 - 4 years       MORTGAGE   
15    16000.0  36 months     7.90       500.65    < 2 years           RENT   
16    11000.0  36 months     7.90       344.20  2 - 4 years       MORTGAGE   
17    20000.0  60 months    19.03       324.47  5 - 9 years           RENT   
18    10700.0  36 months    12.42       357.55  2 - 4 years           RENT   
19     8000.0  36 months     7.90       250.33  2 - 4 years           RENT   
20    30000.0  60 months    19.03       778.72    < 2 years           RENT   
21     6400.0  36 months     8.90       203.23  5 - 9 years            OWN   
22     2250.0  36 months    13.49        76.35  5 - 9 years           RENT   
23    22000.0  60 months    19.42       575.79  2 - 4 years           RENT   
24    18000.0  60 months    15.27       329.07  2 - 4 years           RENT   
25     7175.0  36 months     7.90       224.51  2 - 4 years           RENT   
26     4800.0  36 months     6.03       146.10  5 - 9 years       MORTGAGE   
27    12000.0  60 months    11.71       265.18  2 - 4 years       MORTGAGE   
28    20000.0  60 months    12.42       449.15  2 - 4 years       MORTGAGE   
29    35000.0  60 months    11.71       773.44    10+ years       MORTGAGE   
30     5600.0  36 months     7.51       174.23    10+ years           RENT   
31     3000.0  36 months    10.65        97.72  2 - 4 years           RENT   
32    25000.0  36 months    15.27       869.95  5 - 9 years           RENT   
33    17000.0  36 months    12.42       568.06    < 2 years       MORTGAGE   
34     5000.0  36 months    12.69       167.73  2 - 4 years           RENT   
35     6000.0  36 months     9.91       193.35    < 2 years           RENT   
36     2500.0  60 months    12.69        56.49    < 2 years            OWN   
37    20000.0  60 months    19.91       528.88  5 - 9 years           RENT   
38    16000.0  36 months     7.90       500.65  5 - 9 years           RENT   
39     7100.0  36 months     7.90       222.17    < 2 years           RENT   
40    25000.0  60 months    17.58       629.14    < 2 years       MORTGAGE   
41     9800.0  36 months    12.69       328.74    10+ years       MORTGAGE   
42     3300.0  36 months    14.65       113.84    < 2 years           RENT   
43    10000.0  36 months    11.71       330.76  2 - 4 years           RENT   
44    18725.0  60 months    17.27       468.09  5 - 9 years       MORTGAGE   
45    20000.0  60 months    15.27       478.64  5 - 9 years       MORTGAGE   
46     5000.0  36 months    17.58       179.72    < 2 years           RENT   
47    12100.0  60 months    17.27       302.48  2 - 4 years           RENT   
48    15000.0  60 months    15.96       364.46  5 - 9 years       MORTGAGE   
49    11500.0  36 months     7.90       359.84  2 - 4 years       MORTGAGE   

    annual_inc verificati

Removing the "fico_range" variable as this has now been replaced by "credit_score" showing scores from "Poor" to "Exceptional".

In [62]:
loans_data_clean = loans_data_clean.drop(["fico_range"], axis = 1)
loans_data_clean.head(50)

loan_amnt       term int_rate  installment   emp_length home_ownership  \
0      5000.0  36 months    10.65       162.87    10+ years           RENT   
1      5000.0  36 months     7.90       156.46  2 - 4 years           RENT   
2      6500.0  60 months    14.65       153.45  5 - 9 years            OWN   
3      6000.0  36 months    12.42       200.50    < 2 years           RENT   
4     12000.0  36 months    16.29       423.61  5 - 9 years           RENT   
5      6000.0  36 months    13.49       203.59  5 - 9 years           RENT   
6      5600.0  36 months    14.65       193.17    < 2 years       MORTGAGE   
7      7400.0  36 months    13.49       251.09    < 2 years            OWN   
8     24250.0  60 months    23.91       519.76  5 - 9 years       MORTGAGE   
9      2500.0  36 months    11.71        82.69  5 - 9 years       MORTGAGE   
10    12000.0  36 months     8.90       381.04  5 - 9 years       MORTGAGE   
11    18000.0  60 months    11.71       397.77    < 2 years           RENT   
12     7200.0  36 months     6.03       219.14  5 - 9 years       MORTGAGE   
13     4200.0  60 months    17.27       105.00  5 - 9 years           RENT   
14    20000.0  60 months    18.64       514.86  2 - 4 years       MORTGAGE   
15    16000.0  36 months     7.90       500.65    < 2 years           RENT   
16    11000.0  36 months     7.90       344.20  2 - 4 years       MORTGAGE   
17    20000.0  60 months    19.03       324.47  5 - 9 years           RENT   
18    10700.0  36 months    12.42       357.55  2 - 4 years           RENT   
19     8000.0  36 months     7.90       250.33  2 - 4 years           RENT   
20    30000.0  60 months    19.03       778.72    < 2 years           RENT   
21     6400.0  36 months     8.90       203.23  5 - 9 years            OWN   
22     2250.0  36 months    13.49        76.35  5 - 9 years           RENT   
23    22000.0  60 months    19.42       575.79  2 - 4 years           RENT   
24    18000.0  60 months    15.27       329.07  2 - 4 years           RENT   
25     7175.0  36 months     7.90       224.51  2 - 4 years           RENT   
26     4800.0  36 months     6.03       146.10  5 - 9 years       MORTGAGE   
27    12000.0  60 months    11.71       265.18  2 - 4 years       MORTGAGE   
28    20000.0  60 months    12.42       449.15  2 - 4 years       MORTGAGE   
29    35000.0  60 months    11.71       773.44    10+ years       MORTGAGE   
30     5600.0  36 months     7.51       174.23    10+ years           RENT   
31     3000.0  36 months    10.65        97.72  2 - 4 years           RENT   
32    25000.0  36 months    15.27       869.95  5 - 9 years           RENT   
33    17000.0  36 months    12.42       568.06    < 2 years       MORTGAGE   
34     5000.0  36 months    12.69       167.73  2 - 4 years           RENT   
35     6000.0  36 months     9.91       193.35    < 2 years           RENT   
36     2500.0  60 months    12.69        56.49    < 2 years            OWN   
37    20000.0  60 months    19.91       528.88  5 - 9 years           RENT   
38    16000.0  36 months     7.90       500.65  5 - 9 years           RENT   
39     7100.0  36 months     7.90       222.17    < 2 years           RENT   
40    25000.0  60 months    17.58       629.14    < 2 years       MORTGAGE   
41     9800.0  36 months    12.69       328.74    10+ years       MORTGAGE   
42     3300.0  36 months    14.65       113.84    < 2 years           RENT   
43    10000.0  36 months    11.71       330.76  2 - 4 years           RENT   
44    18725.0  60 months    17.27       468.09  5 - 9 years       MORTGAGE   
45    20000.0  60 months    15.27       478.64  5 - 9 years       MORTGAGE   
46     5000.0  36 months    17.58       179.72    < 2 years           RENT   
47    12100.0  60 months    17.27       302.48  2 - 4 years           RENT   
48    15000.0  60 months    15.96       364.46  5 - 9 years       MORTGAGE   
49    11500.0  36 months     7.90       359.84  2 - 4 years       MORTGAGE   

    annual_inc verificati

Writing cleaned data to csv.

Checking "delinq_2yrs" - The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years

In [63]:
loans_data_clean["delinq_2yrs"].value_counts()

0.0     34290
1.0      3222
2.0       664
3.0       215
4.0        61
5.0        21
6.0        10
7.0         4
8.0         2
11.0        1
9.0         1
Name: delinq_2yrs, dtype: int64

Binning values and converting type from integer to categorical.

In [64]:
loans_data_clean["delinq_2yrs"] = loans_data_clean["delinq_2yrs"].astype("category")

In [65]:
def delinq_cat (row):
    if row["delinq_2yrs"] == 0.00:
        return "0"
    if row["delinq_2yrs"] == 1.0:
        return "1"
    return "2+"

loans_data_clean["delinq_2yrs"] = loans_data_clean.apply(lambda row: delinq_cat(row), axis = 1)
loans_data_clean["delinq_2yrs"].value_counts()

0     34290
1      3222
2+      979
Name: delinq_2yrs, dtype: int64

In [66]:
loans_data_clean.to_csv("../clean_data/loans_data.csv", index=False)